File: js/src/defs.ts
export type Vec2 = [number, number]

export type Model = {
  /** Box Width relative to 1 */
  boxWidth: number
  /** Number of rows */
  rows: number
  /** Columns offsets (relative to 1), where column block are represented by `null` (they have width `2*boxWidth`) */
  columns: Array<number|null>
}

export type Grid = {
  rows: number[]
  cols: number[]
}

export type Box = {
  /** Top-left aka translation from origin */
  t: Vec2
  /** Size aka scale */
  s: Vec2
}

export function columnOffsets({ columns, boxWidth }: Model): number[] {
  return columns.map(c => c === null ? boxWidth : c)
}

/** Box X positions (relative to 1) */
export function boxXs({ columns, boxWidth }: Model): number[] {
  const xs: number[] = []
  let dx = 0

  for (const c of columns) {
    if (c === null) {
      xs.push(dx)
      dx += boxWidth
    }
    else
      dx += c
  }

  return xs
}


/** Box size normalized to the grid being `(1, 1)` */
export function boxSize({ rows, boxWidth }: Model): Vec2 {
  return [boxWidth, 1/rows]
}

export const range = (n: number) => new Array(n).fill(0).map((_, i) => i)

function pairsiwe<T>(xs: T[]): [T, T][] {
  const n = xs.length
  const ps: [T, T][] = []
  for (const i of range(n-1))
    ps.push([xs[i], xs[i+1]])
  return ps
}

export function boxPositions(model: Model): Vec2[] {
  const ps: Vec2[] = []
  for (const [x1, x2] of pairsiwe(boxXs(model))) {
    for (const i of range(model.rows)) {
      const r = i/model.rows
      ps.push([x1, r], [x2, r])
    }
  }
  return ps
}

File: py/src/scoresheet_models/defs/defs.py
Vec2 = tuple[float, float]

def column_offsets(cols: list[float|None], box_width: float) -> list[float]:
  offsets = []
  for c in cols:
    offsets.extend([box_width, box_width] if c is None else [c])
  return offsets

def num_blocks(cols: list[float|None]) -> int:
  return len([x for x in cols if x is None])

def block_cols(cols: list[float|None], box_width: float) -> list[float]:
  """x-positions of block cols (relative to the grid width being 1)"""
  xs = []
  dx = 0
  for c in cols:
    if c is None:
      xs.append(dx)
      dx += 2*box_width
    else:
      dx += c
  return xs

def box_positions(block_cols: list[float], rows: int, box_width: float) -> list[Vec2]:
  ps: list[Vec2] = []
  for c in block_cols:
    for i in range(rows):
      r = i/rows
      ps.append([c, r])
      ps.append([c+box_width, r])
  return ps

def box_size(rows: int, box_width: float) -> Vec2:
    """Box size normalized to the grid being `(1, 1)`"""
    return box_width, 1/rows